src/common/rabbitmq/rabbitmq.service.ts

```diff
@@ -1,112 +1,135 @@
+  constructor(private readonly configService: ConfigService) {}
 
-    if (!hostname || !port || !username || !password) {
-      Logger.error(
-        "RabbitMQ is enabled but missing one or more config variables.",
-        "RabbitMQService",
+  getValueFromConfig(key: string): string | number {
+    const configValue = this.configService.get(`rabbitMq.${key}`);
+    if (!configValue)
+      throw new Error(
+        `RabbitMQ is enabled but missing the config variable ${key}.`,
       );
-    } else {
-      this.connectionOptions = {
-        protocol: "amqp",
-        hostname: hostname,
-        port: port,
-        username: username,
-        password: password,
-      };
+    return configValue;
+  }
 
-      amqp.connect(
-        this.connectionOptions,
-        (connectionError: Error, connection: Connection) => {
-          if (connectionError) {
-            Logger.error(
-              "Connection error in RabbitMQService: " +
-                JSON.stringify(connectionError.message),
-              "RabbitMQService",
-            );
-            return;
-          }
-          this.connection = connection;
+  parseConfig(): void {
+    const connectionOptions: Record<string, string | number> = {};
+    RabbitMQService.configKeys.forEach(
+      (configKey) =>
+        (connectionOptions[configKey] = this.getValueFromConfig(configKey)),
+    );
+    connectionOptions.protocol = "amqp";
+    this.connectionOptions = connectionOptions as typeof this.connectionOptions;
+  }
 
-          this.connection.createChannel(
-            (channelError: Error, channel: Channel) => {
-              if (channelError) {
-                Logger.error(
-                  "Channel error in RabbitMQService: " +
-                    JSON.stringify(channelError.message),
-                  "RabbitMQService",
-                );
-                return;
-              }
-              this.channel = channel;
-            },
-          );
-        },
-      );
-    }
+  async onModuleInit(): Promise<void> {
+    Logger.log("Initializing RabbitMQService.", "RabbitMQService");
+    this.parseConfig();
+    await this.connect();
   }
 
-  private connect(queue: string, exchange: string, key: string) {
-    try {
-      this.channel.assertQueue(queue, { durable: true });
-      this.channel.assertExchange(exchange, "topic", {
-        durable: true,
-      });
-      this.channel.bindQueue(queue, exchange, key);
-    } catch (error) {
-      throw new Error(
-        `Could not connect to RabbitMQ queue ${queue} with exchange ${exchange} and key ${key}.`,
-        { cause: error },
-      );
-    }
+  @OnError()
+  private async connect(): Promise<void> {
+    this._error = "Cannot connect to RabbitMQ";
+    this.connection = await amqplibConnect(this.connectionOptions);
+    this._error = "Channel error in RabbitMQService: ";
+    this.channel = await this.connection.createChannel();
```

Comment:
In my view, the `OnError` decorator brings unnecessary complexities. I personally would use `try catch` here  to make it easy to read. But the idea here is to make it DRY I guess.